### PR TITLE
[3.3.5] Core: Fix Spell Raise Ally

### LINKED ISSUE #82: [DK] Raise Ally
1.Party up with someone
2.Let him die
3.Cast Raise Ally on him

What is the expected output? What do you see instead?
Dead friendly grouped target is supposed to raise as a kewl and strong 
ghoul for 5 minutes under HIS control. Instead, nothing happens. The skill 
goes in Cooldown, but the target remains dead and nothing really happens.

http://www.wowhead.com/spell=61999
Raises the corpse of a raid or party member to fight by your side.  The 
player will have control over the Ghoul for 5 min.

fix Mangos
https://github.com/rsa/mangos/commit/7cbb2fcb3fc0b51139a2cad557ba2d58e4b5d6df
https://github.com/rsa/scriptdev2/commit/ff1db75e8716b04b7aee8b5da807a84919b43569

for part 2
https://github.com/rsa/mangos/commit/96ec786b4071ce1ce6b9415e07aa5c4facd91b11

for part 3
https://github.com/rsa/mangos/commit/3e89e2633b68e39dae9a174d6c7b93e8373ae305

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/66479-dk-raise-ally?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ccrs**
Trying to fix #82

Help will be much appreciated >.<

### CODE REVIEW

**Comment by tkrokli** [comment]
Thanks for starting on this, I was close to PR @Runico's hack from 2013 to get a discussion started.

**Comment by ccrs** [comment]
Travis Build Failed:

````note
The command "sudo apt-get -qq install libssl-dev libmysqlclient-dev libmysql++-dev libreadline6-dev zlib1g-dev libbz2-dev" failed and exited with 100 during .
````

wtf? xd

**Comment by tkrokli** [comment]
Reason:
`dpkg: error: dpkg status database is locked by another process`
`E: Sub-process /usr/bin/dpkg returned an error code (2)`

Try updating your PR locally, then push it again to your repository, thereby forcing Travis to start again.

**Comment by Naios** [comment]
Or you restart the travis build manually...

File: sql/updates/world/2015_XX_XX_XX_world.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+UPDATE `creature_template` SET `minlevel`='80', `maxlevel`='80' WHERE `entry`='30230';
```
Comment (tkrokli) [inline sql/updates/world/2015_XX_XX_XX_world.sql]: Remove the ' ' characters from the numbers, then push to repo and see if Travis will run again.
```sql
UPDATE `creature_template` SET `minlevel`=80, `maxlevel`=80 WHERE `entry`=30230;

File: sql/updates/world/2015_XX_XX_XX_world.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+DELETE FROM `spell_script_names` WHERE `spell_id` IN ();
+INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
+('61999', 'spell_dk_raise_ally_initial'),
+('46619', 'spell_dk_raise_ally');
```
Comment (tkrokli) [inline sql/updates/world/2015_XX_XX_XX_world.sql]:
```sql
(61999, 'spell_dk_raise_ally_initial'),  
(46619, 'spell_dk_raise_ally');

File: sql/updates/world/2015_XX_XX_XX_world.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+UPDATE `creature_template` SET `minlevel`='80', `maxlevel`='80' WHERE `entry`='30230';
+
+DELETE FROM `spell_script_names` WHERE `spell_id` IN ();
```
Comment (ForesterDev) [inline sql/updates/world/2015_XX_XX_XX_world.sql]:
```
DELETE FROM `spell_script_names` WHERE `spell_id` IN (61999,46619);
```

Comment (Shauren) [inline sql/updates/world/2015_XX_XX_XX_world.sql]: dont delete by spell_id, delete by ScriptName

File: sql/updates/world/2015_XX_XX_XX_world.sql (hunk 2 above)
Comment (tkrokli) [inline sql/updates/world/2015_XX_XX_XX_world.sql]: Looking at the suggestions so far, I think the SQL should look something like this in full :
```sql
UPDATE `creature_template` SET `minlevel`=80, `maxlevel`=80 WHERE `entry`=30230;
DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_dk_raise_ally_initial', 'spell_dk_raise_ally');
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
(61999, 'spell_dk_raise_ally_initial'),  
(46619, 'spell_dk_raise_ally');
```
At least this way, the queries are repeatable without causing errors upon a second run.

**Comment by ccrs** [comment]
@MitchesD which spells?

**Comment by ccrs** [comment]
````sql 
UPDATE `creature_template` SET `PowerType` = 3, `spell1` = 62225, `spell2` = 47480, `spell3` = 47481, `spell4` = 47482, `spell5` = 47484, `spell6` = 67866, `ScriptName` = 'npc_risen_ally' WHERE `entry` = 30230;
````
Can any dev confirm this? :)

**Comment by ccrs** [comment]
Right now Ghoul is set to the player which casts the spell.

Modifications on max/minlevel seem unnecessary as its set in
````c++
void TempSummon::InitStats(uint32 duration)
````

**Comment by ccrs** [comment]
Right now im trying to guess where the "set ghoul state to dead player" should be done.

Spell  61999 has these effects
````note
Effect 0: Id 3 (SPELL_EFFECT_DUMMY)
BasePoints = 46619
Targets (57, 0) (TARGET_UNIT_TARGET_RAID, NO_TARGET)

Effect 1: Id 77 (SPELL_EFFECT_SCRIPT_EFFECT)
BasePoints = 1
Targets (18, 8) (TARGET_DEST_CASTER, TARGET_UNIT_DEST_AREA_ENTRY)
Radius (Id 12) 100,00
````

Should be handled on "Effect 1: Id 77 (SPELL_EFFECT_SCRIPT_EFFECT)"?

**Comment by MitchesD** [comment]
I did some research long time ago, will try to find it. There are some auras which have to be scripted and applied on the ghoul, like some modifiers according to player stats etc.

File: src/server/game/Entities/Unit/Unit.h [hunk 4]
```diff
@@ -466,7 +466,8 @@ enum DeathState
     JUST_DIED      = 1,
     CORPSE         = 2,
     DEAD           = 3,
-    JUST_RESPAWNED = 4
+    JUST_RESPAWNED = 4,
+    GHOULED        = 5
```
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.h]: haaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaack

File: src/server/game/Entities/Player/Player.cpp [hunk 5]
```diff
@@ -5038,6 +5038,33 @@ void Player::ResurrectPlayer(float restore_percent, bool applySickness)
...
+    data << uint8(0);
+    data << uint32(0);
+    target->GetSession()->SendPacket(&data);
+}
+
+bool Player::IsValidGhoulResurrectRequest(ObjectGuid guid)
+{
+    if (!m_ghoulResurrectGUID.IsEmpty() && m_ghoulResurrectGUID == guid)
+    {
+        this->CastSpell(this, 46619 /*SPELL_DK_RAISE_ALLY*/, true, nullptr, nullptr, m_ghoulResurrectGUID);
```
Comment (ccrs) [inline src/server/game/Entities/Player/Player.cpp]: Using ObjectGuid originalCaster to take the reference for the whisper,
makes it cast "Effect 0: Id 28 (SPELL_EFFECT_SUMMON)".

Some other way to send that guid to spellscript must be used (or try a different approach).

**Comment by ccrs** [comment]
Current TODO list:

- Hardcoded pet scaling or spellcripts.
- "ObjectGuid originalCaster" issue (wrong SPELL_EFFECT_SUMMON target)
- CorpseReclaimDelay frame doesnt disappear on GhoulResurrect accept.
- Resurrection text is displayed for all group members in chat (wrong range?)
- Test ResurrectionRequest sent having already a GhoulResurrect.
- Unaura handling on despawn/dead (player and creature)

**Comment by Aokromes** [comment]
We can’t automatically merge this pull request.
Use the command line to resolve conflicts before continuing.

**Comment by tkrokli** [comment]
Good job so far, glad to see that the PR still passes Travis build after the squash. Nice ToDo list as well.

**Comment by tkrokli** [comment]
@ccrs : looks like Travis is acting up again. You may want to try squashing your commits to restart Travis.
```
pycurl.error: (7, "couldn't connect to host")
The command "echo "yes" | sudo add-apt-repository ppa:kalakris/cmake" failed and exited with 1 during .
Your build has been stopped.

**Comment by ccrs** [comment]
Current TODO list:

- Hardcoded pet scaling or spellcripts.
- CorpseReclaimDelay frame doesnt disappear on GhoulResurrect accept.
- Test ResurrectionRequest sent having already a GhoulResurrect.

**Comment by ariel-** [comment]
You can use a CreatureAI for the ghoul with member

```c++
void OnCharmed(bool /*apply*/) override { }
```

So that your AI doesn't get replaced by a CharmedAI and you can handle the spells there.

**Comment by ccrs** [comment]
OnCharmed is just a hook, cant prevent the replace of the AI.

**Comment by ariel-** [comment]
It can, if you don't override this hook the virtual default implementation gets used, which is

```c++
//Disable CreatureAI when charmed
void CreatureAI::OnCharmed(bool /*apply*/)
{
    //me->IsAIEnabled = !apply;*/
    me->NeedChangeAI = true;
    me->IsAIEnabled = false;
}
```

Source: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/CreatureAI.cpp#L31-36

**Comment by ccrs** [comment]
When charm is removed its called on next update tick, by then ghoul is already removed from world.

File: src/server/scripts/Spells/spell_dk.cpp [hunk 6]
```diff
@@ -1689,6 +1691,234 @@ class spell_dk_death_grip_initial : public SpellScriptLoader
...
+        PrepareSpellScript(spell_dk_raise_ally_initial_SpellScript);
+
+        bool Validate(SpellInfo const* /*spellInfo*/) override
+        {
+            if (!sSpellMgr->GetSpellInfo(SPELL_DK_RAISE_ALLY_INITIAL))
+                return false;
+            return true;
+        }
+
+        SpellCastResult CheckCast()
```
Comment (MitchesD) [inline src/server/scripts/Spells/spell_dk.cpp]: I think you should check cast for player targets and also spell shoudn't be used in arena. Try this:
``cpp
            SpellCastResult CheckTarget()
            {
                if (Unit* target = GetTargetUnit())
                {
                    if (target->GetTypeId() != TYPEID_PLAYER || target->isAlive())
                        return SPELL_FAILED_BAD_TARGETS;
                    else if (GetCaster()->ToPlayer()->InArena())
                        return SPELL_FAILED_NOT_IN_ARENA;

                    return SPELL_CAST_OK;
                }
            }
``

Comment (ccrs) [inline src/server/scripts/Spells/spell_dk.cpp]: i'll check if it can be cast con arena, but player part isn't necessary as it has attributes to only target players

Comment (robinsch) [inline src/server/scripts/Spells/spell_dk.cpp]: I'm not familar with the spell script system, one small question. Does the CheckTarget() overrides the Spell::CheckTarget?

Comment (Shauren) [inline src/server/scripts/Spells/spell_dk.cpp]: No, it allows you to write additional conditions

File: src/server/game/Entities/Player/Player.cpp [hunk 7]
```diff
@@ -5038,6 +5041,35 @@ void Player::ResurrectPlayer(float restore_percent, bool applySickness)
...
+    data << uint64(GetGUID());
+    data << uint32(0);
+    data << 0;
+    data << uint8(0);
+    data << uint8(0);
+    data << uint32(0);
+    target->GetSession()->SendPacket(&data);
+}
+
+bool Player::IsValidGhoulResurrectRequest(ObjectGuid guid)
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: Getter methods (Get/Is...) are very bad places to modify the object (Player)

Comment (ccrs) [inline src/server/game/Entities/Player/Player.cpp]: ok, ill add an extra function.

Comment (StormBytePP) [inline src/server/game/Entities/Player/Player.cpp]: I tested this, and have 2 little problems (not sure how was in retail at that time);
1) You can't talk when you're dead, shows when you try to /s
2) "Release spirit" still appears on screen

As I said, I am not sure about retail (nor I don't have sniffs), but I think that you should be able to talk at least, and the release spirit message not to appear

**Comment by StormBytePP** [comment]
Also, nonPCH fails to compile, can you apply:
```diff
diff --git a/src/server/scripts/Spells/spell_dk.cpp b/src/server/scripts/Spells/spell_dk.cpp
index 51f24d6..15ff9f0 100644
--- a/src/server/scripts/Spells/spell_dk.cpp
+++ b/src/server/scripts/Spells/spell_dk.cpp
@@ -22,6 +22,7 @@
  */
 
 #include "Player.h"
+#include "UnitAI.h"
 #include "ScriptMgr.h"
 #include "SpellScript.h"
 #include "SpellAuraEffects.h"
```

**Comment by StormBytePP** [comment]
@ccrs Ping? Can you try to do the proposed changes?

**Comment by kvipka** [comment]
+ need 
https://github.com/healthstone/RE/commit/7c66e704e5ee6539fa12f1cda2d22a04e55bbf5b

and 
https://github.com/healthstone/RE/commit/5f5e5325432980a9dc32dc69e2a873dfdb693007

**Comment by ccrs** [comment]
I already knew about that, thx

**Comment by Aokromes** [comment]
@ccrs https://github.com/TrinityCore/TrinityCore/pull/14830#issuecomment-134145852

**Comment by Runico** [comment]
@ccrs, is this usable without problem ?

**Comment by tkrokli** [comment]
@Runico : notice the suggestion from StormBytePP and Aokromes to add https://github.com/TrinityCore/TrinityCore/pull/14830#issuecomment-134145852
```diff
diff --git a/src/server/scripts/Spells/spell_dk.cpp b/src/server/scripts/Spells/spell_dk.cpp
index 51f24d6..15ff9f0 100644
--- a/src/server/scripts/Spells/spell_dk.cpp
+++ b/src/server/scripts/Spells/spell_dk.cpp
@@ -22,6 +22,7 @@
  */

 #include "Player.h"
+#include "UnitAI.h"
 #include "ScriptMgr.h"
 #include "SpellScript.h"
 #include "SpellAuraEffects.h"
```
(new line to be added after existing line 24, before existing line 25)
https://github.com/TrinityCore/TrinityCore/pull/14830/files#diff-f55cd9074f1ab13e4acd486d33268828R25

File: src/server/game/Entities/Player/Player.cpp [hunk 8]
```diff
@@ -5038,6 +5041,44 @@ void Player::ResurrectPlayer(float restore_percent, bool applySickness)
...
+{
+    if (!m_ghoulResurrectPlayerGUID.IsEmpty() && m_ghoulResurrectPlayerGUID == guid)
+        return true;
+
+    return false;
+}
+
+void Player::GhoulResurrect()
+{
+    this->CastSpell(this, 46619 /*SPELL_DK_RAISE_ALLY*/, true, nullptr, nullptr, m_ghoulResurrectPlayerGUID);
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.cpp]: "this->" is usually not used around TC sources

File: src/server/game/Spells/Spell.cpp [hunk 9]
```diff
@@ -4192,17 +4192,14 @@ void Spell::SendResurrectRequest(Player* target)
 {
     // get resurrector name for creature resurrections, otherwise packet will be not accepted
     // for player resurrections the name is looked up by guid
-    std::string const sentName(m_caster->GetTypeId() == TYPEID_PLAYER
-                               ? ""
-                               : m_caster->GetNameForLocaleIdx(target->GetSession()->GetSessionDbLocaleIndex()));
+    std::string const sentName(m_caster->GetTypeId() == TYPEID_PLAYER ?
```
Comment (jackpoz) [inline src/server/game/Spells/Spell.cpp]: these changes seems unrelated to the PR, please avoid applying codestyle changes mixed with functional changes

File: src/server/scripts/Spells/spell_dk.cpp [hunk 10]
```diff
@@ -1689,6 +1692,298 @@ class spell_dk_death_grip_initial : public SpellScriptLoader
...
+            summon->CastSpell(summon, SPELL_DK_DEATH_KNIGHT_RUNE_WEAPON_SCALING_02, true);
+            summon->CastSpell(summon, SPELL_DK_DEATH_KNIGHT_PET_SCALING_01, true);
+            summon->CastSpell(summon, SPELL_DK_DEATH_KNIGHT_PET_SCALING_03, true);
+            summon->CastSpell(summon, SPELL_DK_PET_SCALING___MASTER_SPELL_03___INTELLECT_SPIRIT_RESILIENCE, true);
+            summon->CastSpell(summon, SPELL_DK_PET_SCALING___MASTER_SPELL_06___SPELL_HIT_EXPERTISE_SPELL_PENETRATION, true);
+
+            // SMSG_POWER_UPDATE is sent
+            summon->SetMaxPower(POWER_ENERGY, 100);
+
+            if (GetCaster()->ToPlayer())
```
Comment (jackpoz) [inline src/server/scripts/Spells/spell_dk.cpp]: please store the ToPlayer() result in a local variable in the if( ) check to avoid doing the same cast twice, triggering static analysis issues

File: src/server/scripts/Spells/spell_dk.cpp [hunk 11]
```diff
@@ -1689,6 +1692,298 @@ class spell_dk_death_grip_initial : public SpellScriptLoader
...
+                return false;
+            return true;
+        }
+
+        void CalcDamage()
+        {
+            if (Aura* aur = GetCaster()->GetAura(SPELL_GHOUL_FRENZY))
+            {
+                int32 damage = GetHitDamage();
+                damage += int32(GetCaster()->GetTotalAttackPowerValue(BASE_ATTACK) * 0.05 * aur->GetStackAmount());
```
Comment (jackpoz) [inline src/server/scripts/Spells/spell_dk.cpp]: is 0.05 double intentionally ?

**Comment by jackpoz** [comment]
@StormBytePP I see you self-assigned this PR, what's the current status ?

**Comment by Shauren** [comment]
I have been notified that this PR causes crashes, worth investigating

**Comment by Takenbacon** [comment]
Logout while transformed to a ghoul. ( ͡° ͜ʖ ͡°)

**Comment by jackpoz** [comment]
can you post a proper crash report ?

**Comment by ccrs** [comment]
yes, there are some issues related.

File: src/server/game/Spells/Spell.cpp (hunk 9 above)
Comment (ccrs) [inline src/server/game/Spells/Spell.cpp]:
````
extra byte on SMSG_RESURRECT_REQUEST spotted by @Robens`
````

**Comment by jackpoz** [comment]
were the issues related to this PR addressed ?

**Comment by ccrs** [comment]
yes

**Comment by jackpoz** [comment]
What about the crash and "Logout while transformed to a ghoul" ?

**Comment by tkrokli** [comment]
Based on what StormByePP said on one of his own PRs about having too much to do IRL to follow up, I think it would be relevant to assign this PR to someone else (or remove assignment).

**Comment by ccrs** [comment]
The PR is finished, but there are flaws that might prevent its merge.
* I used a hack to enable stat scaling, detailed on description and code
* There are some parts that are unrelated to the issue
